spec/Section 6 -- Execution.md

```diff
@@ -322,39 +312,450 @@ Unsubscribe(responseStream):
+  - Set the corresponding entry on {incrementalResult} to {incremental}.
+- If {completed} is not empty:
+  - Set the corresponding entry on {incrementalResult} to {completed}.
+- If {pending} is not empty:
+  - Set the corresponding entry on {incrementalResult} to {pending}.
+- Return {incrementalResult}.
+
+GetIncrementalEntry(incrementalDataRecord, graph):
+
+- Let {deferredFragments} be the Deferred Fragments incrementally completed by
+  {incrementalDataRecord} at {path}.
+- Let {result} be the result of {incrementalDataRecord}.
+- Let {data} and {errors} be the corresponding entries on {result}.
+- Let {releasedDeferredFragments} be the members of {deferredFragments} that are
+  root nodes in {graph}.
+- Let {bestDeferredFragment} be the member of {releasedDeferredFragments} with
+  the shortest {path} entry.
+- Let {subPath} be the portion of {path} not contained by the {path} entry of
+  {bestDeferredFragment}.
+- Let {id} be the unique identifier for {bestDeferredFragment}.
+- Return an unordered map containing {id}, {subPath}, {data}, and {errors}.
+
+GetCompletedEntry(pendingResult, errors):
+
+- Let {id} be the unique identifier for {pendingResult}.
+- Let {completedEntry} be an unordered map containing {id}.
+- If {errors} is not empty, set the corresponding entry on {completedEntry} to
+  {errors}.
+- Return {completedEntry}.
+
+### Batching Incremental Results
+
+BatchIncrementalResults(incrementalResults):
+
+- Return a new stream {batchedIncrementalResults} which yields events as
+  follows:
+- While {incrementalResults} is not closed:
+  - Let {availableIncrementalResults} be a list of one or more Incremental
+    Results available on {incrementalResults}.
+  - Let {batchedIncrementalResult} be an unordered map created by merging the
+    items in {availableIncrementalResults} into a single unordered map,
+    concatenating list entries as necessary, and setting {hasNext} to the value
+    of {hasNext} on the final item in the list.
+  - Yield {batchedIncrementalResult}.
+
+### Field Collection
+
+Before execution, the _selection set_ is converted to a grouped field set by
+calling {CollectFields()}. Each entry in the grouped field set is a list of
+fields that share a response key (the alias if defined, otherwise the field
+name). This ensures all fields with the same response key (including those in
+referenced fragments) are executed at the same time.
+
+As an example, collecting the fields of this selection set would collect two
+instances of the field `a` and one of field `b`:
+
+```graphql example
+{
+  a {
+    subfield1
+  }
+  ...ExampleFragment
+}
+
+fragment ExampleFragment on Query {
+  a {
+    subfield2
+  }
+  b
+}
+```
+
+The depth-first-search order of the field groups produced by {CollectFields()}
+is maintained through execution, ensuring that fields appear in the executed
+response in a stable and predictable order.
+
+CollectFields(objectType, selectionSet, variableValues, deferUsage,
```

Comment:
Maybe we should call `deferUsage` => `abstractDeferUsage` and include a note that CollectFields may be memoized?

Reply:
Something like:

Note: When completing a List field, the {CollectFields} algorithm is invoked with the same arguments for each element of the list. GraphQL Services may choose to memoize {CollectFields} for performance.

Reply:
> Maybe we should call `deferUsage` => `abstractDeferUsage` and include a note that CollectFields may be memoized?

I like the note idea but I think renaming to abstractDeferUsage gets a bit confusing, as a DeferUsage is where the directive is actually used in the operation and is always path independent, ie the same no matter if nested under a list within the returned data…. Renaming it to abstractDeferUsage might imply that there is a concreteDeferUsage somewhere….